### PR TITLE
GitHub Issue NOAA-EMC/GSI#89. Update enkf4fv3sar following the EMC GSI master.

### PR DESCRIPTION
**Pull request by TingLei-NOAA**
Update enkf for fv3sar following interface changes ,like, in readgridata, in the enkf for gfs.  
This update only incorporates changes in the interface like including dummy subroutine readgridata_pnc  while without actual implementation of those new functions.  They are to be finished in the future.

### CODE REVIEW

**Comment by MichaelLueken-NOAA** [comment]
Hi Ting,

Please create a new issue for this work.  No changes may enter the master without an issue associated with it.  Additionally, since there are changes to source code files, I'll need to assign a reviewer.  Who would you like to review the two source file changes?

Thanks,

Mike

**Comment by TingLei-NOAA** [comment]
Hi, Mike,
Got it. I will create an issue soon.
I think, Ming Hu or Guoqing Ge are good as the reviewer.
Thanks.
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by TingLei-NOAA** [comment]
Hi, Mike,
Would you confirm that an new issue can only be opened by a project
leader?  Right?
I forgot this for I have some operations on other repository:).
Thanks
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by MichaelLueken-NOAA** [comment]
Hi Ting,

Anyone can open an issue, but I need to assign issues and reviewers.

Mike

**Comment by TingLei-NOAA** [comment]
Mike,
Got it!
So many thanks!
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

File: src/enkf/gridio_fv3reg.f90 [hunk 1]
```diff
@@ -680,6 +682,58 @@ subroutine writegriddata(nanal1,nanal2,vars3d,vars2d,n3d,n2d,levels,ndim,vargrid
     !======================================================================
 
   end subroutine writegriddata
+  subroutine writeincrement(nanal1,nanal2,vars3d,vars2d,n3d,n2d,levels,ndim,grdin,no_inflate_flag)
```
Comment (hu5970) [inline src/enkf/gridio_fv3reg.f90:685]: Please add brief description of this subroutine

File: src/enkf/observer_fv3reg.f90 [hunk 2]
```diff
@@ -62,7 +62,8 @@ subroutine setup_linhx(rlat, rlon, time, ix, delx, ixp, delxp, iy, dely,  &
 
 end subroutine setup_linhx
 
-subroutine calc_linhx(hx, dens, dhx_dx, hx_ens, &
```
Comment (hu5970) [inline src/enkf/observer_fv3reg.f90:65]: Ting, do those changes come from you?

**Comment by hu5970** [comment]
This PR only changed one file in EnKF, right?

File: src/enkf/observer_fv3reg.f90 (hunk 2 above)
Comment (TingLei-NOAA) [reply src/enkf/observer_fv3reg.f90:65]: Yes, I think this is to make it has the same form as the definition in observer_gfs.f90  and the one  called in the readconvobs.f90 and son on.

File: src/enkf/gridio_fv3reg.f90 (hunk 1 above)
Comment (TingLei-NOAA) [reply src/enkf/gridio_fv3reg.f90:685]: Done!

**Comment by TingLei-NOAA** [comment]
According to the display on website, there were files changed.

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by hu5970** [comment]
@TingLei-NOAA 

The code in EnKF looks good. 

There are other code changes in gsi directory. Do you want to review that part of code too. Or should the code in GSI directory be a separate PR?

**Comment by TingLei-NOAA** [comment]
Hi, Ming,
Thanks for pointing that out.  That should be modifications for update of
delz, which Xiaoyan is in charge of its PR.
I thought the original plan was Xiaozhan would do that PR separately.
Hi, Mike,
Would you clarify on what the status is for those delz related changes in
the gsi code?
I know there were some actions that Xiaoyan's stuff was committed to my
master fork ,which was for the current PR.
what really happened is kind of confusing for me:).
So, your clarification is really appreciated.
I will also cc Xiaoyan in this thread.
Regards,
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by MichaelLueken-NOAA** [comment]
Hi Ting,

There are a few issues with your latest push, so I'll give instructions to correct:

1) git clone --recursive git@github.com:TingLei-NOAA/GSI.git update
2) cd update
3) git reset --hard c6058837223b46bc3e13eb640875034f63120944

Now, to merge the latest master in NOAA-EMC/GSI into your fork (from the NOAA-EMC/GSI GitHub wiki):

1) git remote add upstream https://github.com/NOAA-EMC/GSI.git
2) git remote update
3) git rebase upstream/master
Your fork will now be up-to-date with the NOAA-EMC/GSI master, with your commit at the HEAD of your fork's master.

Now, apply the changes that Ming noted, then use:

1) git add {modified files}
2) git commit --amend (please don't change current commit message)
3) git push origin master --force

If you have any questions, please let me know.

Mike

**Comment by TingLei-NOAA** [comment]
Hi, Mike,
Seems "git clone --recursive git@github.com:TingLei-NOAA/GSI.git update"
doen't work for me now,
It gave error message
^^
" ssh: connect to host github.com port 29418: Connection timed out
fatal: Could not read from remote repository.
Please make sure you have the correct access rights
and the repository exists.
VVV
At the same, the command like "git clone --recursive
https://github.com/NOAA-EMC/GSI.git " works.
Would you please help with this issue (another form if possible)?
Thanks
Ting



______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by MichaelLueken-NOAA** [comment]
Hi Ting,

Replace:

git clone --recursive git@github.com:TingLei-NOAA/GSI.git update

with

git clone --recursive https://github.com/TingLei-NOAA/GSI.git update

Mike

**Comment by TingLei-NOAA** [comment]
Hi, Mike,
Thanks a lot.
Now I have finished all those steps you gave.
Hope that will be the PR we planned to be ready:).
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by MichaelLueken-NOAA** [comment]
Hi Ting,

It looks like everything is back to the way it should be.

Thanks,

Mike

**Comment by TingLei-NOAA** [comment]
Mike,
Great to know everything is ok now:).
Thanks.
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Review by hu5970 — APPROVED** [review]

**Comment by MichaelLueken-NOAA** [comment]
Hi Ting,

I have completed my review of your changes.  The regression tests were run and all passed successfully.  The code was then compiled in debug mode, with no issues noted.  The debug tests were run and all passed as well.  With respect to the visual inspection of the changes, I see no issues.  Having said that, you will need to remove the fix modification from your PR.  It appears as though the update will revert quite of bit of fix file changes.  So, it will need to be removed.  To remove the fix update:

1) git clone --recursive git@github.com:TingLei-NOAA/GSI.git update
2) cd update
3) git reset --soft HEAD~1
4) git reset HEAD fix
5) git commit -m "update enkf4fv3sar following the EMC GSI master"
6) git push origin master --force

Once these steps have been taken, I should be able to get these changes out to the review committee.

Thanks,

Mike

**Comment by TingLei-NOAA** [comment]
Hi, Mike,
Not sure how my actions caused that problem with fix.
Now, I had finished the steps you gave (except for the first step , for
which I used the address https://github.com/TingLei-NOAA/GSI.git).
Hope this is what you need.
Regards,
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by MichaelLueken-NOAA** [comment]
The due date has past with now feedback from the review committee, so I will give final approval to these changes and merge them to the NOAA-EMC/GSI master.

**Pull request merged**